### PR TITLE
[pnnl/SHAD#18] Added SSSP-length algorithm, vertex nomination example…

### PR DESCRIPTION
**Pull request by VitoCastellana**
…, and fixed a race-condition bug in the catalog

### CODE REVIEW

File: examples/edge_index_graph/ei-sssp.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  shad::EdgeIndex<size_t, size_t>::ObjectID OID(-1);
+  auto loadingTime = shad::measure<std::chrono::microseconds>::duration([&]() {
+    // The GraphReader expects an input file in METIS dump format
+    std::ifstream inputFile;
+    inputFile.open(argv[1], std::ifstream::in);
+    OID = GraphReader(inputFile);
+  });
+
+  std::cout << "Graph loaded in " << loadingTime.count() / (double)1000000
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-sssp.cc]: fixed updating measure utility

File: examples/edge_index_graph/ei-sssp.cc [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  size_t target = std::stoul(argv[3], nullptr, 0);
+  auto duration = shad::measure<std::chrono::microseconds>::duration([&]() {
+    path_length =
+        sssp_length<shad::EdgeIndex<size_t, size_t>, size_t>(OID, src, target);
+  });
+
+  if (path_length != std::numeric_limits<size_t>::max()) {
+    std::cout << "Found a path between " << src << " and " << target << " in "
+              << path_length << " hops in "
+              << duration.count() / (double)1000000 << " seconds" << std::endl;
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-sssp.cc]: fixed updating measure utility

File: examples/edge_index_graph/ei-sssp.cc [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  } else {
+    std::cout << "Couldn't find a path between " << src << " and " << target
+              << " in " << duration.count() / (double)1000000 << " seconds"
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-sssp.cc]: fixed updating measure utility

File: examples/edge_index_graph/ei-vertex_nomination.cc [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+//===----------------------------------------------------------------------===//
+
+// Simple implementation of triangle counting, through graph pattern matching
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-vertex_nomination.cc]: fixed

File: examples/edge_index_graph/ei-vertex_nomination.cc [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            << std::endl;
+  auto eiPtr = shad::EdgeIndex<size_t, size_t>::GetPtr(OID);
+
+  size_t num_vertices = eiPtr->Size();
+  std::cout << "NumVertices: " << num_vertices
+            << " Num Edges: " << eiPtr->NumEdges() << std::endl;
+
+  // call srdan on all localities
+  auto callSrand = [](const size_t &seed) {
+    srand(seed + static_cast<uint32_t>(rt::thisLocality()));
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-vertex_nomination.cc]: moving this to a separate issue

File: examples/edge_index_graph/ei-vertex_nomination.cc [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+  };
+  rt::executeOnAll(callSrand, seed);
+
+  for (size_t i = 0; i < num_runs; ++i) {
+    // create seeds
+    // using a vector instead of a set to facilitate serialization
+    std::vector<size_t> rand_seeds;
+    size_t rnd_seed = 0;
+    while (rand_seeds.size() < num_seeds) {
+      rnd_seed = rand() % num_vertices + 1;
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-vertex_nomination.cc]: fixed

File: examples/edge_index_graph/ei-vertex_nomination.h [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+//===----------------------------------------------------------------------===//
+
+// Simple implementation of triangle counting, through graph pattern matching
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-vertex_nomination.h]: fixed

File: examples/edge_index_graph/ei-vertex_nomination.h [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  size_t path_length = min_path;
+  for (auto seed : seeds) {
+    if (node_id == seed) continue;
+    // sssp from node_id to seed for better locality
+    path_length = sssp_length<shad::EdgeIndex<size_t, size_t>, size_t>(
+        GID, node_id, seed);
+    min_path = std::min<size_t>(min_path, path_length);
+  }
+
+  return 1.0 / static_cast<float>(min_path);
```
[the remark this answers is not in the document]
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-vertex_nomination.h]: fixed

File: examples/edge_index_graph/ei-sssp.cc [hunk 9]
```diff
@@ -27,6 +27,7 @@
 #include <atomic>
 #include <fstream>
 #include <iostream>
+#include <ratio>
```
Comment (mminutoli) [inline examples/edge_index_graph/ei-sssp.cc]: Is this needed?

File: include/shad/util/measure.h [hunk 10]
```diff
@@ -56,13 +56,14 @@ struct measure {
...
-    auto duration = std::chrono::duration_cast<TimeT>(
-        std::chrono::steady_clock::now() - start);
+    std::chrono::duration<double, typename TimeT::period> duration =
+                   std::chrono::steady_clock::now() - start;
 
     return duration;
   }
 };
 
+
```
Comment (mminutoli) [inline include/shad/util/measure.h]: Can you remove this extra space?  Did you run make clang-format?

File: examples/edge_index_graph/ei-sssp.cc (hunk 9 above)
Comment (VitoCastellana) [reply examples/edge_index_graph/ei-sssp.cc]: not anymore

File: include/shad/util/measure.h (hunk 10 above)
Comment (VitoCastellana) [reply include/shad/util/measure.h]: done

**Pull request merged**